### ISSUE TITLE
eslint throws error on arrow function class methods

### ISSUE THREAD

**Issue by codeocelot** [eslint/eslint#4683]
I think I've found a bug.  According to the es6 spec, class methods are free to use arrow function declarations.  However, eslint throws a `Parsing error: unexpected token =` when it encounters these methods.  

Try the following: 
```js
    class App{
        classMethod = () => { console.log('hi world');}
    }
```
I've tried configuring my .eslintrc to cover all the obvious features without success.  
```
    {
      "ecmaFeatures": {
        "modules":true,
        "arrowFunctions":true,
        "classes":true
      },
      "env": {
        "es6": true
      },
    }
```

Edit: I'm using v1.10.3

**Comment by eslintbot** [comment]
@codeocelot Thanks for the issue! If you're reporting a bug, please be sure to include:

1. The version of ESLint you are using (run `eslint -v`)
2. What you did (the source code and ESLint configuration)
3. The actual ESLint output complete with numbers
4. What you expected to happen instead

Requesting a new rule? Please see [Proposing a New Rule](http://eslint.org/docs/developer-guide/contributing/new-rules) for instructions.

**Comment by gyandeeps** [comment]
I havent read anywhere that we can have arrow function like that inside classes. FYI My ES6 knowledge is very limited. 
Can you share some link which explains this?

**Comment by mysticatea** [comment]
@codeocelot ES6 does not support class properties. You cannot use arrow functions as class method.

See Also: http://www.ecma-international.org/ecma-262/6.0/#sec-class-definitions

**Comment by ilyavolodin** [comment]
As per comments above, arrow functions are not allowed as class properties in ES6. Feel free to reopen if you find information that states otherwise.

**Comment by codeocelot** [comment]
In general, there is very little documentation about if arrow class methods are allowed.  However, there are numerous examples of it being done.  In fact, React's [own docs](https://facebook.github.io/react/docs/reusable-components.html#no-autobinding) suggest using it to overcome the lack of autobinding `this`.  

Unfortunately, [MDN's own 'Arrow Function'](https://developer.mozilla.org/en/docs/Web/JavaScript/Reference/Functions/Arrow_functions) doc doesn't discuss classes at all.

**Comment by ilyavolodin** [comment]
@codeocelot React also recommends using object rest/spread which is not part of the ES6 spec. I'm not up to date on ES6 features, but @mysticatea linked official specification, which, as far as I understand, says that arrow functions are not allowed as class properties. I believe this is a proposed feature for ES7.

**Comment by mysticatea** [comment]
@codeocelot 

The spec of Class Property Declarations is stage 1 proposal currently: https://github.com/tc39/ecma262#current-proposals
See Also: https://tc39.github.io/process-document/

This feature is not part of ES2015(ES6) and perhaps not part of ES2016(ES7) also.
So ESLint does not support it yet.

`babel-eslint` might help you to use features which are under discussion.
Please note the spec might be changed in future.

**Comment by dekelb** [comment]
Following @mysticatea's comment, if you are using babel and you still want eslint to work based on that, you can add

    "parser": "babel-eslint",

to your `eslintrc` file.

**Comment by michaelaflores** [comment]
For somer reason this no longer works for me. Eslint is now throwing an error that a class method as arrow function was used before defined. Anyone else with a similar setup facing this issue?

.eslintrc
`{
  "extends": "eslint-config-airbnb",
  "env": {
    "browser": true,
    "node": true,
    "mocha": true
  },
  "rules": {
    "camelcase": 0,
    "comma-dangle": 0,
    "guard-for-in": 1,
    "import/default": 0,
    "import/extensions": 0,
    "import/no-duplicates": 0,
    "import/no-extraneous-dependencies": 0,
    "import/named": 0,
    "import/namespace": 0,
    "import/no-unresolved": 0,
    "import/no-named-as-default": 2,
    "indent": [2, 2, {
      "SwitchCase": 1
    }],
    "jsx-a11y/interactive-supports-focus": 1,
    "max-len": 0,
    "new-cap": [2, {
      "capIsNewExceptions": ["List", "Map", "Set"]
    }],
    "no-alert": 1,
    "no-console": 2,
    "no-unused-vars": 2,
    "no-use-before-define": ["error", {
      "functions": false
    }],
    "react/jsx-boolean-value": 0,
    "react/jsx-filename-extension": 0,
    "react/no-multi-comp": 0,
    "global-require": 0
  },
  "parser": "babel-eslint",
  "parserOptions": {
    "ecmaFeatures": {
      "jsx": true
    }
  },
  "plugins": [
    "react", "import"
  ],
  "settings": {
    "import/parser": "babel-eslint",
    "import/resolve": {
      "moduleDirectory": ["node_modules", "src"]
    }
  },
  "globals": {
    "__DEVELOPMENT__": true,
    "__CLIENT__": true,
    "__SERVER__": true,
    "__DISABLE_SSR__": true,
    "__DEVTOOLS__": true,
    "webpackIsomorphicTools": true
  }
}`

Eslint version
`   "eslint": "^3.19.0",
    "eslint-config-airbnb": "^15.0.1",
    "eslint-plugin-import": "2.2.0",
    "eslint-plugin-jsx-a11y": "5.0.1",
    "eslint-plugin-react": "7.0.1",`

**Comment by stantoncbradley** [comment]
broke upgrading `eslint`/`airbnb-eslint`. fixed for us by upgrading `babel-eslint`

**Comment by gogumai** [comment]
Thanks @stantoncbradley that worked for me!

**Comment by Amurmurmur** [comment]
@stantoncbradley @gogumai A problem for me. eslint throws an error that I have to define whenever I write something like:
```
toggleModal = () => {
    window.alert("Test")
  }
```

My eslint config:
```
{
    "parser": "babel-eslint",
    "env": {
        "es6": true,
        "node": true
    },
    "extends": [
        "eslint:recommended",
        "plugin:react/recommended",
        "standard",
        "react-native",
        "airbnb"
    ],
    "parserOptions": {
        "ecmaFeatures": {
            "experimentalObjectRestSpread": true,
            "jsx": true,
            "sourceType": "module",
            "allowImportExportEverywhere": false
        },
        "sourceType": "module"
    },
    "plugins": [
        "react",
        "react-native",
        "redux-saga",
        "jsx-a11y"
    ],
    "rules": {
        "indent": [
            "error",
            2
        ],
        "linebreak-style": [
            "error",
            "unix"
        ],
        "quotes": [
            "error",
            "single"
        ],
        "semi": [
            "error",
            "never"
        ],
        "comma-dangle": 1,
        "import/imports-first": 0,
        "import/first": 0,
        "import/newline-after-import": 0,
        "import/no-dynamic-require": 0,
        "import/no-extraneous-dependencies": 0,
        "import/no-named-as-default": 0,
        "import/no-named-as-default-member": 1,
        "import/no-unresolved": 2,
        "import/prefer-default-export": 0,
        "react/jsx-uses-react": "error",
        "react/jsx-uses-vars": "error",
        "jsx-a11y/aria-props": 2,
        "jsx-a11y/heading-has-content": 0,
        "jsx-a11y/href-no-hash": 0,
        "jsx-a11y/label-has-for": 2,
        "jsx-a11y/mouse-events-have-key-events": 2,
        "jsx-a11y/role-has-required-aria-props": 2,
        "jsx-a11y/role-supports-aria-props": 2,
        "max-len": 0,
        "generator-star-spacing": 0,
        "space-before-function-paren": 0,
        "newline-per-chained-call": 0,
        "no-confusing-arrow": 0,
        "no-console": 1,
        "no-use-before-define": 0,
        "no-underscore-dangle": 0,
        "prefer-template": 2,
        "spaced-comment": 1,
        "react/forbid-prop-types": 0,
        "react/jsx-first-prop-new-line": [
            2,
            "multiline"
        ],
        "react/jsx-filename-extension": 0,
        "react/jsx-no-target-blank": 0,
        "react/require-default-props": 0,
        "react/require-extension": 0,
        "react/self-closing-comp": 0,
        "redux-saga/no-yield-in-race": 2,
        "redux-saga/yield-effects": 2,
        "require-yield": 0,
        "react/jsx-no-bind": [
            2,
            {
                "ignoreRefs": false,
                "allowArrowFunctions": false,
                "allowBind": false
            }
        ]
    }
}
```
RN v0.52
Node v9.3.0

**Comment by drewdrewthis** [comment]
I'm also still having this issue.

**Comment by not-an-aardvark** [comment]
Please create a new issue and fill out the [bug report template](https://raw.githubusercontent.com/eslint/eslint/master/templates/bug-report.md) if you're having trouble.
